### PR TITLE
Enable cross-multipage dfn.js

### PR DESCRIPTION
**Pull request by sideshowbarker**
This change enables the dfn popups in multipage output to show links to all
cross-references for a <dfn> term across all files in the multipage output,
using a JSON data file generated by wattsi.

For single-page output this change replaces the existing dfn.js mechanism with
the same one used for multipage output (in the interest of not needing to
maintain separate mechanisms for single-page vs multipage for the HTML spec).

But since dfn.js is used by other specs, we can’t replace it, so we’ll neither
to either:

1. Move this multipage dfn.js to whatwg/html-build (since at this point, the
   HTML spec is the only spec we have multipage output for anyway).

2. Or, rename this dfn-multipage.js and keep it here.

In the mean time, the diff currently in this PR enables this change to be
reviewed by comparing the new code here to the code it’s replacing.

Depends on https://github.com/whatwg/wattsi/pull/46

### CODE REVIEW

**Comment by zcorpan** [comment]
I like option 2 better; it seems plausible for bikeshed-generated specs to at some point support multipage as well. Or even without multipage, it might be beneficial to load the references from a JSON file on request instead of generating the data from the DOM on load.

**Comment by zcorpan** [comment]
Well... actually, bikeshed has its own dfn-panel implementation.
https://github.com/tabatkins/bikeshed/blob/5bd73bb15eb04ad9f7d1a57f012e9ee6eca5a765/bikeshed/__init__.py#L1566

So maybe we could have all non-HTML specs switch to that, and then make this thing be specific for HTML.

**Comment by domenic** [comment]
I tend to go for option 1 myself as this will be pretty HTML-specific.

File: dfn.js [hunk 1]
```diff
@@ -92,73 +68,54 @@ function dfnShow(event) {
...
-          a.onclick = dfnMovePanel;
-          if (header != lastHeader) {
-            lastHeader = header;
-            n = 1;
-            var li = document.createElement('li');
-            if (typeof header == 'string') {
-              a.appendChild(document.createTextNode(header));
+            if (i === 0) {
+              var headerFormatted = header.replace(/</g, '&lt;');
+              headerFormatted = headerFormatted.replace(/ ([^ ]+) (element(?!s)|attribute(?!s)|interface(?!s)|common interface|object)/g, ' <code>$1</code> $2');
```
Comment (domenic) [inline dfn.js]: Where did all this come from? Was it just not as good before and this helps fix it?

Comment (zcorpan) [inline dfn.js]: In the old implementation it copied the full DOM except for `a` and `dfn` elements. In the new the JSON just contains the textContent: https://github.com/whatwg/wattsi/pull/46/files#diff-1922eb90770bdd34ac96cc9db339faebR605

The new approach seems like it will "break" when we add some heading with `<code>` that doesn't match this pattern. Would it be possible to roundtrip some code-markers in the JSON?

Comment (sideshowbarker) [inline dfn.js]: > Where did all this come from? Was it just not as good before and this helps fix it?

That second `headerFormatted` replacement that’s adding the `<code>…</code>` around particular words is there because we don’t preserve the markup within the headings when we store it (in the `xref.json` file).

So if we don’t do that replacement to add the `<code>…</code>` back for those cases, then we lose the formatting around those words in the dfn popops. Which would be a regression in UX.

Most of the rest of the `headerFormatted` replacements are there to clean up after some false-positive cases that the other replacement wrongly causes to get marked up with `<code>…</code>`.

And that all wasn’t necessary previously because, before, the old implementation looked only at the DOM of the current document, and as @zcorpan notes, it copied the DOM. So all the markup was preserved. But making it work across multiple documents requires writing the headings in some form out to some persistent store. So since we’re writing it out to a JSON file,  in wattsi we take just the textContent and save that.

Comment (sideshowbarker) [inline dfn.js]: > The new approach seems like it will "break" when we add some heading with `<code>` that doesn't match 
this pattern. Would it be possible to roundtrip some code-markers in the JSON?

Of course it’s possible and as you might imagine, I already considered doing that. But consider what the requires  doing in wattsi: instead of just taking the textContent, we need to instead write a serializer and walk through all the heading contents in that wattsi code, and serialize the markup along with the text.

It doesn’t seem to me worth it to do that extra work in wattsi—in Pascal code none of us working on this currently are very familiar with, and that takes us a lot more time to write and debug—since we can instead do it in the JavaScript in the way I implemented it here, and it handles all the current headings in the spec without breaking.

And yeah I guess if we late use `<code>…</code>` for some uncommon case that the current JavaScript isn’t catching, then we will not get the `<code>…</code>` preserved and it won’t show up with code formatting in the dfn popups. But I wouldn’t exactly see that as “breaking”  (and I guess you wouldn’t either since you also put “break” in quotes…). Instead I would see it as a relatively minor degrade in UX for an isolated case.

In other words it’s a tradeoff for trying to keep the wattsi code no more complex than what we can reasonably get by with. But that doesn’t prejudice us from later deciding we want to handle the markup serializing in the wattsi code after, if it’s clearly better to do it that way. It just doesn’t seem clear at this point yet that it is.

Comment (zcorpan) [inline dfn.js]: That's fair enough, and I don't object to landing this workaround, though it's tempting to take a stab the serializer idea...

The format could be an array where every other item is a code string, e.g.:
```
header: ["The ", "lang", " and ", "xml:lang", " attributes"],
```

Comment (domenic) [inline dfn.js]: I would have thought Wattsi already contained a serializer, since it has to write out the output anyway.

But I am also happy to merge any intermediate fix here. The only blocker in my mind for this whole series of patches is https://github.com/whatwg/html/pull/2638#issuecomment-300162981 since it is a regression, if I understand correctly.

Comment (sideshowbarker) [inline dfn.js]: > I would have thought Wattsi already contained a serializer, since it has to write out the output anyway.

As far as I can see it doesn’t have something that’s exposed in a way that’d enable what we want here. But it’s also still possible I’m just missing something.

> The only blocker in my mind for this whole series of patches is whatwg/html#2638 (comment) since it is a regression, if I understand correctly.

Yes, and I plan to get that fixed in the next couple days

**Comment by sideshowbarker** [comment]
> But I am also happy to merge any intermediate fix here. The only blocker in my mind for this whole series of patches is whatwg/html#2638 (comment) since it is a regression

Yes. But I’ve pushed 8151c92 as a fix for that just now, and https://sideshowbarker.net/html/multipage/ and https://sideshowbarker.net/html/ have the latest

File: dfn.js (hunk 1 above)
Comment (zcorpan) [inline dfn.js]: Just an existence proof that the current hack doesn't cover everything, as motivation for better roundtripping (i.e., not to change the regexp): see \#the-fieldset-and-legend-elements

Comment (sideshowbarker) [inline dfn.js]: > Just an existence proof that the current hack doesn't cover everything, as motivation for better roundtripping (i.e., not to change the regexp): see #the-fieldset-and-legend-elements

Thanks yeah so there are probably more of those than I imagined but still think however many, it seems like a reasonable tradeoff for now to put up with some number of those, and just fix them (by updated the regex replacements) as we come across them.

But for the long term I agree it at this point that it makes sense to plan to update the wattsi code to deal this more systematically/robustly.

**Comment by domenic** [comment]
So this looks good, but we should as discussed move it to whatwg/html as a html-specific dfn.js. (Maybe "html-dfn.js"?)

**Comment by domenic** [comment]
Moved to https://github.com/whatwg/html/pull/2638 in preparation for merge. This was the right way to do a review though, as a delta from dfn.js, so yay.
